### PR TITLE
Export part interfaces instead of constructors

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: packages/lit-html/CHANGELOG.md [hunk 1]
```diff
@@ -78,6 +78,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
   and `PartInfo` types have been moved from `lit-html` to
   `lit-html/directive.ts`.
 
+- [Breaking] (since 2.0.0-pre.3) The part exports (`ChildPart`,
+  `AttributePart`, etc) have been change to interface-only exports. The constructors are no longer exported.
```
Comment (kevinpschaaf) [inline packages/lit-html/CHANGELOG.md:82]:
```suggestion
  `AttributePart`, etc) have been change to interface-only exports. The constructors are no longer exported. Directive authors should use helpers in `directive-helpers.js` to construct parts.
```

File: packages/lit-html/src/private-ssr-support.ts [hunk 2]
```diff
@@ -52,4 +52,10 @@ export const _$private = {
     return committedValue;
   },
   resolveDirective: p._resolveDirective,
+  // ChildPart: p._ChildPart,
+  // AttributePart: p._AttributePart,
+  // PropertyPart: p._PropertyPart,
+  // BooleanAttributePart: p._BooleanAttributePart,
+  // EventPart: p._EventPart,
```
Comment (kevinpschaaf) [inline packages/lit-html/src/private-ssr-support.ts:59]: Remove?

File: packages/lit-html/src/test/lit-html_test.ts [hunk 3]
```diff
@@ -12,10 +12,11 @@
  * http://polymer.github.io/PATENTS.txt
  */
 import {
+  _$private,
```
Comment (kevinpschaaf) [inline packages/lit-html/src/test/lit-html_test.ts:15]: Remove?

File: packages/lit-html/src/platform-support.ts [hunk 4]
```diff
@@ -87,11 +87,11 @@ const scopeCssStore: Map<string, string[]> = new Map();
...
 (globalThis as any)['litHtmlPlatformSupport'] ??= ({
-  ChildPart,
-  Template,
+  _$ChildPart: ChildPart,
+  _$Template: Template,
 }: {
-  ChildPart: PatchableChildPart;
-  Template: PatchableTemplate;
+  _$ChildPart: PatchableChildPart;
+  _$Template: PatchableTemplate;
```
Comment (kevinpschaaf) [inline packages/lit-html/src/platform-support.ts:94]: I would just change these to an argument list; doesn't seem worth burning more stable properties to make this an options bag...

**Review by kevinpschaaf — CHANGES_REQUESTED** [review]

File: packages/lit-html/CHANGELOG.md (hunk 1 above)
Comment (justinfagnani) [reply packages/lit-html/CHANGELOG.md]: done

File: packages/lit-html/src/private-ssr-support.ts (hunk 2 above)
Comment (justinfagnani) [reply packages/lit-html/src/private-ssr-support.ts]: done

File: packages/lit-html/src/platform-support.ts (hunk 4 above)
Comment (justinfagnani) [reply packages/lit-html/src/platform-support.ts]: done

cc @sorvell to confirm. It means future changes just have to be adding to the list, but maybe we don't expect any.

File: packages/lit-html/src/test/lit-html_test.ts (hunk 3 above)
Comment (justinfagnani) [reply packages/lit-html/src/test/lit-html_test.ts]: done

File: packages/lit-html/src/lit-html.ts [hunk 5]
```diff
@@ -819,7 +829,10 @@ export type Part =
   | ElementPart
   | EventPart;
 
-export class ChildPart {
+type Interface<T> = {[P in keyof T]: T[P]};
+
+export type ChildPart = Interface<ChildPartImpl>;
+class ChildPartImpl implements ChildPart {
```
Comment (sorvell) [inline packages/lit-html/src/lit-html.ts:835]: Is the `implements` necessary? Isn't this just "I implement myself"?

File: packages/lit-html/src/lit-html.ts [hunk 6]
```diff
@@ -1063,7 +1076,8 @@ export class ChildPart {
   }
 }
 
-export class AttributePart {
+export type AttributePart = Interface<AttributePartImpl>;
+class AttributePartImpl {
```
Comment (sorvell) [inline packages/lit-html/src/lit-html.ts:1080]: Why doesn't this have the `implements` thing?

**Review by sorvell — APPROVED** [review]
LGTM, just a couple nits

File: packages/lit-html/src/lit-html.ts (hunk 5 above)
Comment (justinfagnani) [reply packages/lit-html/src/lit-html.ts:835]: oh oops, this is from when I had a hand-written interface

**Review by kevinpschaaf — APPROVED** [review]

**Pull request merged**